backend/src/crd/controller/viewer/reconciler/reconciler.go

```diff
@@ -0,0 +1,224 @@
+	glog.Infof("Reconcile request: %+v", req)
+
+	view := &viewerV1alpha1.Viewer{}
+	if err := r.Get(context.Background(), req.NamespacedName, view); err != nil {
+		if errors.IsNotFound(err) {
+			// No viewer instance, so this may be the result of a delete.
+			// Nothing to do.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	glog.Infof("Got instance: %+v", view)
+
+	// Ignore other viewer types for now.
+	if view.Spec.Type != viewerV1alpha1.ViewerTypeTensorboard {
+		glog.Infof("Unsupported spec type: %q", view.Spec.Type)
+		// Return nil to indicate nothing more to do here.
+		return reconcile.Result{}, nil
+	}
+
+	// Set up potential deployment.
+	dpl, err := deploymentFrom(view)
+	if err != nil {
+		utilruntime.HandleError(err)
+		// User error, don't requeue key.
+		return reconcile.Result{}, nil
+	}
+
+	// Set the deployment to be owned by the view instance. This ensures that
+	// deleting the viewer instance will delete the deployment as well.
+	if err := controllerutil.SetControllerReference(view, dpl, r.scheme); err != nil {
+		// Error means that the deployment is already owned by some other instance.
+		utilruntime.HandleError(err)
+		return reconcile.Result{}, err
+	}
+
+	foundDpl := &appsv1.Deployment{}
+	nsn := types.NamespacedName{Name: dpl.Name, Namespace: dpl.Namespace}
+	if err := r.Client.Get(context.Background(), nsn, foundDpl); err != nil {
+		if errors.IsNotFound(err) {
+			// Create a new instance.
+			if err := r.Client.Create(context.Background(), dpl); err != nil {
+				utilruntime.HandleError(fmt.Errorf("error creating deployment: %v", err))
+				return reconcile.Result{}, err
+			}
+		} else {
+			// Some other error.
+			utilruntime.HandleError(err)
+			return reconcile.Result{}, err
+		}
+	}
+	glog.Infof("Created new deployment with spec: %+v", dpl)
+
+	// Set up a service for the deployment above.
+	svc := serviceFrom(view, dpl.Name)
+	// Set the service to be owned by the view instance as well.
+	if err := controllerutil.SetControllerReference(view, svc, r.scheme); err != nil {
+		// Error means that the service is already owned by some other instance.
+		utilruntime.HandleError(err)
+		return reconcile.Result{}, err
+	}
+
+	foundSvc := &corev1.Service{}
+	nsn = types.NamespacedName{Name: svc.Name, Namespace: svc.Namespace}
+	if err := r.Client.Get(context.Background(), nsn, foundSvc); err != nil {
+		if errors.IsNotFound(err) {
+			// Create a new instance.
+			if err := r.Client.Create(context.Background(), svc); err != nil {
+				utilruntime.HandleError(fmt.Errorf("error creating service: %v", err))
+				return reconcile.Result{}, err
+			}
+		} else {
+			// Some other error.
+			utilruntime.HandleError(err)
+			return reconcile.Result{}, err
+		}
+	}
+	glog.Infof("Created new service with spec: %+v", svc)
+
+	return reconcile.Result{}, nil
+}
+
+func setPodSpecForTensorboard(view *viewerV1alpha1.Viewer, s *corev1.PodSpec) {
```

Comment:
Are you setting the owner references for the created resources so that garbage collection is taken care of? 

Do you need labels so that we can query by type of resource? Otherwise the UI may list all the resources which could be a problem once we add additional types.